sim/priest/talents.go

```diff
@@ -60,9 +61,644 @@ func (priest *Priest) ApplyTalents() {
+			return
+		}
+
+		priest.AddMana(sim, priest.MaxMana()*0.05*float64(priest.Talents.Masochism), manaMetrics)
+	}
+
+	priest.RegisterAura(
+		core.Aura{
+			Label:    "Masochism",
+			Duration: core.NeverExpires,
+			OnReset: func(aura *core.Aura, sim *core.Simulation) {
+				aura.Activate(sim)
+			},
+			OnSpellHitTaken: func(_ *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+				if !result.Landed() {
+					return
+				}
+
+				damageTakenHandler(sim, result.Damage)
+			},
+
+			OnPeriodicDamageTaken: func(_ *core.Aura, sim *core.Simulation, _ *core.Spell, result *core.SpellResult) {
+				damageTakenHandler(sim, result.Damage)
+			},
+
+			OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+				// SW:D
+				if priest.ShadowWordDeath == spell && result.Landed() {
+					priest.AddMana(sim, priest.MaxMana()*0.05*float64(priest.Talents.Masochism), manaMetrics)
+					return
+				}
+			},
+		},
+	)
+}
+
+func (priest *Priest) applyMindMelt() {
+	if priest.Talents.MindMelt == 0 {
+		return
+	}
+
+	mindMeltMod := priest.AddDynamicMod(core.SpellModConfig{
+		ClassMask:  int64(PriestSpellMindBlast),
+		FloatValue: -0.5,
+		Kind:       core.SpellMod_CastTime_Pct,
+	})
+
+	priest.MindMeltProcAura = priest.RegisterAura(core.Aura{
+		Label:     "Mind Melt Proc",
+		ActionID:  core.ActionID{SpellID: 87160},
+		Duration:  time.Second * 15,
+		MaxStacks: 2,
+		OnStacksChange: func(_ *core.Aura, _ *core.Simulation, oldStacks int32, newStacks int32) {
+			mindMeltMod.UpdateFloatValue(-0.5 * float64(newStacks))
+		},
+
+		OnExpire: func(_ *core.Aura, _ *core.Simulation) {
+			mindMeltMod.Deactivate()
+		},
+
+		OnCastComplete: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell) {
+			if priest.MindBlast == spell {
+				aura.Deactivate(sim)
+			}
+		},
+	})
+
+	priest.RegisterAura(core.Aura{
+		Label:    "Mind Melt",
+		Duration: core.NeverExpires,
+		OnReset: func(aura *core.Aura, sim *core.Simulation) {
+			aura.Activate(sim)
+		},
+
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			if result.Landed() && priest.MindSpike == spell {
+				priest.MindMeltProcAura.Activate(sim)
+				priest.MindMeltProcAura.AddStack(sim)
+			}
+		},
+	})
+}
+
+func (priest *Priest) applySinAndPunishment() {
+	if priest.Talents.SinAndPunishment == 0 {
+		return
+	}
+
+	priest.RegisterAura(core.Aura{
+		Label:    "SinAndPunishment",
+		Duration: core.NeverExpires,
+		OnReset: func(aura *core.Aura, sim *core.Simulation) {
+			aura.Activate(sim)
+		},
+
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			// check for critical mind flay tick
+			if result.Outcome.Matches(core.OutcomeCrit) && priest.MindFlayAPL.ClassSpellMask == int64(PriestSpellMindFlay) {
+
+				// reduce cooldown
+				remaining := max(0, time.Duration(*priest.Shadowfiend.CD.Timer)-time.Second*5)
```

Comment:
`CD.TimeToReady(sim)` or `CD.ReadyAt()` depending on what you need (You can check the other available methods you should not need to cast this to anything)